skyd/cache.go

```diff
@@ -0,0 +1,168 @@
+		result:   nil,
+		skylinks: make(map[string]struct{}),
+		mu:       sync.Mutex{},
+	}
+}
+
+// Add registers the given skylinks in the cache.
+func (psc *PinnedSkylinksCache) Add(skylinks ...string) {
+	psc.mu.Lock()
+	defer psc.mu.Unlock()
+	for _, s := range skylinks {
+		psc.skylinks[s] = struct{}{}
+	}
+}
+
+// Contains returns true when the given skylink is in the cache.
+func (psc *PinnedSkylinksCache) Contains(skylink string) bool {
+	psc.mu.Lock()
+	defer psc.mu.Unlock()
+	_, exists := psc.skylinks[skylink]
+	return exists
+}
+
+// Diff returns two lists of skylinks - the ones that are in the given list but
+// are not in the cache (missing) and the ones that are in the cache but are not
+// in the given list (removed).
+func (psc *PinnedSkylinksCache) Diff(sls []string) (unknown []string, missing []string) {
+	psc.mu.Lock()
+	defer psc.mu.Unlock()
+	removedMap := make(map[string]struct{}, len(psc.skylinks))
+	for sl := range psc.skylinks {
+		removedMap[sl] = struct{}{}
+	}
+	for _, sl := range sls {
+		// Remove this skylink from the removedMap, because it has not been
+		// removed.
+		delete(removedMap, sl)
+		// If it's not in the cache - add it to the added list.
+		_, exists := psc.skylinks[sl]
+		if !exists {
+			unknown = append(unknown, sl)
+		}
+	}
+	// Transform the removed map into a list.
+	for sl := range removedMap {
+		missing = append(missing, sl)
+	}
+	return
+}
+
+// Rebuild rebuilds the cache of skylinks pinned by the local skyd. The
+// rebuilding happens in a goroutine, allowing the method to return a channel
+// on which the caller can either wait or select. The caller can check whether
+// the rebuild was successful by calling Error().
+func (psc *PinnedSkylinksCache) Rebuild(skydClient NodeSkydClient) RebuildCacheResult {
+	psc.mu.Lock()
+	defer psc.mu.Unlock()
+	if !psc.isRebuildInProgress() {
+		psc.result = &RebuildCacheResult{
+			Ch:        make(chan struct{}),
+			ExternErr: nil,
+		}
+		// Kick off the actual rebuild in a separate goroutine.
+		go psc.threadedRebuild(skydClient)
+	}
+	return *psc.result
+}
+
+// Remove registers the given skylinks in the cache.
+func (psc *PinnedSkylinksCache) Remove(skylinks ...string) {
+	psc.mu.Lock()
+	defer psc.mu.Unlock()
+	for _, s := range skylinks {
+		delete(psc.skylinks, s)
+	}
+}
+
+// isRebuildInProgress returns true if a cache rebuild is in progress.
+// Calling this method assumes that caller is holding a lock on the cache.
+func (psc *PinnedSkylinksCache) isRebuildInProgress() bool {
+	return psc.result != nil
+}
+
+// threadedRebuild performs the actual cache rebuild process. It reports any
+// errors by setting the psc.err variable and it always closes the rebuildCh on
+// exit.
+func (psc *PinnedSkylinksCache) threadedRebuild(skydClient NodeSkydClient) {
+	var err error
+	// Ensure that we properly wrap up the rebuild process.
+	defer func() {
+		psc.mu.Lock()
+		// Update the result.
+		psc.result.ExternErr = err
+		close(psc.result.Ch)
```

Comment:
Follow Up: for these concurrent channels, we have implemented `Close()` functions on the struct in `skyd` so that we can check if the channel was already closed before closing it. This helps protect against panics for closing a closed channel.

Reply:
We have this implemented only for files, it seems, and we typically just call `close`, e.g.
```go
close(uc.staticUploadCompletedChan)
```
But if you'd prefer that, I added a `close` method which checks whether the channel is closed before closing it, thus preventing double-closing. The method is unexported because I don't want it to acquire its own locks - I want the setting of the error and the closing of the channel to happen in that specific order and within the same lock scope.